src/metabase/api/dashboard.clj

```diff
@@ -215,14 +216,97 @@
       hide-unreadable-cards
       add-query-average-durations))
 
+(defn- cards-to-copy
+  "Returns a map of which cards we need to copy and which are not to be copied. The `:copy` key is a map from id to
+  card. The `:discard` key is a vector of cards which were not copied due to permissions."
+  [ordered-cards]
+  (letfn [(split-cards [{:keys [card series] :as db-card}]
+            (cond
+              (nil? (:card_id db-card)) ;; text card
+              []
+
+              (mi/can-write? card)
+              (let [{writable true unwritable false} (group-by (comp boolean mi/can-write?)
+                                                               series)]
+                [(into [card] writable) unwritable])
+              ;; if you can't write the base, we don't have anywhere to put the series
+              :else
+              [[] (into [card] series)]))]
+    (reduce (fn [acc db-card]
+              (let [[retain discard] (split-cards db-card)]
+                (-> acc
+                    (update :copy merge (m/index-by :id retain))
+                    (update :discard concat discard))))
+            {:copy {}
+             :discard []}
+            ordered-cards)))
+
+(defn- duplicate-cards
+  "Takes a dashboard id, and duplicates the cards both on the dashboard's cards and dashcardseries. Returns a map of
+  {old-card-id duplicated-card} so that the new dashboard can adjust accordingly."
+  [dashboard dest-coll-id]
+  (let [same-collection? (= (:collection_id dashboard) dest-coll-id)
+        {:keys [copy discard]} (cards-to-copy (:ordered_cards dashboard))]
+    (reduce (fn [m [id card]]
+              (assoc m id
+                     (if (:dataset card)
```

Comment:
could you explain why we don't duplicate the model?

Reply:
Consider models to be tables. They are more permanent that just cards. They also might be persisted to the source database and we just copy a reference to them.